vortex-serde/src/layouts/write/layouts.rs

```diff
@@ -1,100 +1,66 @@
-use std::collections::VecDeque;
-
+use bytes::Bytes;
 use flatbuffers::{FlatBufferBuilder, WIPOffset};
 use vortex_flatbuffers::{footer as fb, WriteFlatBuffer};
 
-use crate::layouts::LayoutId;
-use crate::stream_writer::ByteRange;
+use crate::layouts::{ChunkedLayoutSpec, ColumnLayoutSpec, FlatLayoutSpec, LayoutId};
 
 #[derive(Debug, Clone)]
-pub enum Layout {
-    Nested(NestedLayout),
-    Flat(FlatLayout),
+pub struct Layout {
+    id: LayoutId,
+    children: Option<Vec<Layout>>,
+    metadata: Option<Bytes>,
 }
 
-impl WriteFlatBuffer for Layout {
-    type Target<'a> = fb::Layout<'a>;
-
-    fn write_flatbuffer<'fb>(
-        &self,
-        fbb: &mut FlatBufferBuilder<'fb>,
-    ) -> WIPOffset<Self::Target<'fb>> {
-        let layout_variant = match self {
-            Self::Nested(l) => l.write_flatbuffer(fbb).as_union_value(),
-            Self::Flat(l) => l.write_flatbuffer(fbb).as_union_value(),
-        };
-
-        let mut layout = fb::LayoutBuilder::new(fbb);
-        layout.add_layout_type(match self {
-            Self::Nested(_) => fb::LayoutVariant::NestedLayout,
-            Self::Flat(_) => fb::LayoutVariant::FlatLayout,
-        });
-        layout.add_layout(layout_variant);
-        layout.finish()
+impl Layout {
+    pub fn flat(begin: u64, end: u64) -> Self {
+        Self {
+            id: FlatLayoutSpec::ID,
+            children: None,
+            metadata: Some(Bytes::copy_from_slice(
+                [begin.to_le_bytes(), end.to_le_bytes()].as_flattened(),
+            )),
+        }
     }
-}
 
-#[derive(Debug, Copy, Clone)]
-pub struct FlatLayout {
-    pub(crate) range: ByteRange,
-}
-
-impl WriteFlatBuffer for FlatLayout {
-    type Target<'a> = fb::FlatLayout<'a>;
-
-    fn write_flatbuffer<'fb>(
-        &self,
-        fbb: &mut FlatBufferBuilder<'fb>,
-    ) -> WIPOffset<Self::Target<'fb>> {
-        fb::FlatLayout::create(
-            fbb,
-            &fb::FlatLayoutArgs {
-                begin: self.range.begin,
-                end: self.range.end,
-            },
-        )
+    pub fn chunked(children: Vec<Layout>, has_metadata: bool) -> Self {
+        Self {
+            id: ChunkedLayoutSpec::ID,
+            children: Some(children),
+            metadata: Some(Bytes::copy_from_slice(&[has_metadata as u8])),
```

Comment:
worth a comment that `has_metadata` is referring to having a metadata array as the first chunk, since otherwise this reads really oddly